### PR TITLE
Use 'io/console', improve JRuby and Windows support, code coverage and add 'acceptance' tests

### PR DESCRIPTION
**Pull request by abinoam**
Hi @JEG2 and all,

This PR brings the beginning of that idea of making some kind of interactive (acceptance) tests to HighLine.
Could you (all) test it with ```rake acceptance``` ?

Another thing I'm excited about is the use of 'io/console'. 
Remember those comments https://github.com/JEG2/highline/issues/85#issuecomment-72475678 and https://github.com/JEG2/highline/issues/85#issuecomment-72754914 ?
I made some test here on my environment and I was able to run HighLine on JRuby using this new HighLine::Terminal::IOConsole mode.

I've put JRuby on our Travis matrix (allowed to fail) so we keep an eye on JRuby compatibility from now on.

Bumped up version to 2.0.0-develop.2

I'll let this PR open for some days to get some feedback (and perhaps add some commits more).

The complete Changelog below.

### 2.0.0-develop.2 / 2015-09-09

(by Abinoam P. Marques Jr. - @abinoam)

#### NOTES

This version brings greater compatibility with JRuby and Windows.
But we still have a lot of small issues in both platforms.
We were able to unify/converge all approaches into using io/console,
so we could delete old code that relied solely on stty, termios, java api and
windows apis (DL and Fiddle).

Another improvement is the beginning of what I called "acceptance tests".
If you type ```rake acceptance``` you'll be guided through some tests
where you have to input some thing and see if everything work as expected.
This makes easier to catch bugs that otherwise would be over-sighted.

#### CHANGES SUMMARY

* Fix Simplecov - it was reporting erroneous code coverage
* Add new tests. Improves code coverage
* Extract HighLine::BuiltinStyles
* Try to avoid nil checking
* Try to avoid class variables (mis)use
* Fix RDoc include path and some small fixes to the docs
* Move HighLine::String to its own file
* Add HighLine::Terminal::IOConsole
  - Add an IOConsoleCompatibility module with some stubbed
    methods for using at StringIO, File and Tempfile to help
    on tests.
  - Any enviroment that can require 'io/console' will
    use HighLine::Terminal::IOConsole by default. This kind
    of unifies most environments where HighLine runs. For
    example, we can use Terminal::IOConsole on JRuby!!!
* Add ruby-head and JRuby (19mode and head) to Travis CI matrix. Yes, this
  our first step to a more peaceful JRuby compatibility.
* Add AppVeyor Continuous Integration for Windows
* Add _acceptance_ tests for HighLine
  - Use ```rake acceptance``` to run them
  - Basically it interactively asks the user to confirm if
    some expected HighLine behavior is actually happening.
    After that it gather some environment debug information,
    so the use could send to the HighLine contributors in case
    of failure.
* Remove old and unused files (as a result of relying on io/console)
  - JRuby
  - Windows (DL and Fiddle)
  - Termios
* Fix some small (old and new) bugs
* Make some more tuning for Windows compatibility
* Make some more tuning for JRuby compatibility

ANY feedback is really appreciated!

### CODE REVIEW

**Comment by abinoam** [comment]
After I have upgrade Java to 1.8 and JRuby to 9.0.0.0 I'm having a issue with readline.
Reported on JRuby github repo https://github.com/jruby/jruby/issues/3181

**Comment by djberg96** [comment]
+1 for this patch. Looking for JRuby support myself.

**Comment by abinoam** [comment]
Hi @djberg96 ,

I'm having some problems that was not able to test in different environments by now.
But I think I'll have time for fixing it next week and merge this PR.
Thanks for your feedback.

**Comment by djberg96** [comment]
@abinoam Anything I can help with? Which environments?

**Comment by abinoam** [comment]
@djberg96 yes, here the tests are passing ok too.

BUT... (automatically) testing something that rely on console interactions sometimes doesn't reflect the actual behaviour. This happens because most tests uses StringIO to simulate the actual console. And StringIO doesn't have all "properties" of an actual console.

That's why I'm trying to introduce some tests that "really" depends on user interaction.
I have called it here "acceptance" tests, although I'm not sure if it applies when done "semi-manually".

One thing you could do to me is to clone this branch and run "rake acceptance" and copy/paste its output here at the issue. Could you do this with JRuby in your enviroment? It will run the scripts from test/acceptance

Perhaps you could also install this alpha release with ```rake install``` and run the scripts directly.

The problem I'm having is with JRuby readline is...
When HighLine uses (JRuby) Readline autocomplete feature, after gathering the input, the terminal settings gets misconfigured and it doesn't accepts <enter> any more, and instead it prints out __^M__

I can easily "fix" this just by saving stty configurations and restoring it after using Readline. 
You can look at this diff https://github.com/abinoam/highline/compare/feat_windows_compat...abinoam:fix_readline_at_jruby

But I'm finding this situation a little strange as I haven't noticed this before upgrading Java and JRuby 9.0.0.0 and I would like to understand it better.

```
=====
   at_readline.rb
=====

This step checks if the readline autocomplete feature is working.
The test has 5 options you can choose from: save, sample, exec, exit and load.
...
Don't forget to answer 'y' (yes) or 'n' (no) to the question at the end.
---
Enter command:  ex
exec   exit
Enter command:  exit
Executing "exit"...
---

Did the Readline autocomplete work fine (y/n)? y^M^M^M
```

Look at the "^M^M^M" it's me trying to keypress <enter>. You'll have to ctrl-C to exit.

**Comment by djberg96** [comment]
@abinoam yep, I see the same thing. I'll try to take a look at it. Maybe you could hop into #jruby on freenode, they might be able to help.

**Comment by djberg96** [comment]
@abinoam, just out of curiosity, how attached are you to using readline? Would a pure Ruby solution like rawline perhaps be easier to work with?

**Comment by djberg96** [comment]
@abinoam I did notice some places where you're matching against 1.7 but perhaps that should be >= 1.7.0 ? So far it hasn't fixed the acceptance tests yet, though. Just thought I'd mention it.

**Comment by abinoam** [comment]
I didn't tried the IRC channel.
But I have open an issue on JRuby. ( jruby/jruby#3181) No answers till now. The devs should be really really busy with this 9.0.0.0 release.

@djberg96 Thanks for pointing me to https://github.com/h3rald/rawline I was not aware of it.
Yes, I'm open to changes, specially in this alpha state for 2.0.0 we're currently in.
And I'm sure the boss (@JEG2) is open too.
You can see that we're going toward relying more on "io/console" than on "stty" tricks because we have perceived that now io/console is on standard library and is "more" available and reliable.
Readline is the choice because is the "mainstream" solution for line completion.
But we can go toward https://github.com/h3rald/rawline but we have to be sure that it fits our needs.
Meanwhile, the fastest path is to make readline work.

**Comment by abinoam** [comment]
I'm on IRC now.

**Comment by djberg96** [comment]
@abinoam, I see that it's ultimately a bug in jline2 that your workaround fixes, so I'm hoping this can get merged.

The rawline gem does have a readline compatibility mode, but I couldn't make it work. Still, something to consider for the future.

Thanks for your work!

**Comment by abinoam** [comment]
Hi @djberg96 thanks for your feedback.
I was a little busy this week.
I'm planning to do some manual testing tomorrow so I could merge (confidently).

**Comment by abinoam** [comment]
Hi @JEG2 (cc @djberg96)

I think we are ready to ship 2.0.0-develop.2 version.

Could you do a quick review at the commits so we could merge it?

Is the code going in the right direction?

**Comment by abinoam** [comment]
Just for keep a record. I have made some tests with RbReadline as substitute to builtin readline so we could be "free" of calling ```stty``` in   Readline context. It worked, but I have had some issues with multi-byte characters, so I have to investigate it futher.

And Rawline is looking unmaintained.

**Comment by abinoam** [comment]
I have left saved my experimentation with RbReadline on https://github.com/abinoam/highline/tree/rbr-3

**Comment by djberg96** [comment]
@abinoam Please report your issue at https://github.com/ConnorAtherton/rb-readline

**Comment by abinoam** [comment]
@djberg96 I will try to.
The first issue was solved and it was related to the way rb-readline loads itself.
The other one I will try to isolate and write a simple sample test script.

**Comment by djberg96** [comment]
@abinoam Thanks.

Shame about Rawline. Would you be interested in picking up maintenance? The author might be looking for someone to take over.

**Comment by abinoam** [comment]
Rawline is a great gem, but I stilIl have much work to do with HighLine until I have a good 2.0.0 production version to ship. I think I'm going slowly here and this is because of lack of time. So I prefer to do one thing better than burn out trying to save the world! Perhaps when things get more "calm" here I can take another gem to contribute.

**Comment by abinoam** [comment]
@djberg96 could test this gist on your setup?

https://gist.github.com/abinoam/c8e81f7c0e851bee40f4

**Comment by djberg96** [comment]
@abinoam I put a comment on the gist. Definitely something wrong. I didn't realize Luis had turned over maintenance of rb-readline to someone else. There seem to be a lot of open issues.

I also noticed that there's at least one fork of Rawline, by Zach Dennis, that seems to be getting updates. I've contacted him about taking over the project.

**Comment by JEG2** [comment]
I just read through the commits.

This continues to be the most epically great refactoring effort I have ever seen.  We must find ways to share this with the world:  blog posts, conference talks, a book, or whatever.  This is amazing stuff!

Almost every commit I see something else that I really love:

* Class variables are gone
* Examples and tests now show what mode HighLine is in
* IO/console replaces most of the scary low-level code

Hooray!

Super minor comments:

* Feel free to convert my horrible old habit of abusing `and` and `or` in conditionals to the correct `&&` and `||`.
* Have you checked if control-C works when `io/console` is waiting on `getch()`?  Just curious.

:+1: x 1,000

Thank you so, so much!!!

**Comment by abinoam** [comment]
:blush:

Thank you very much for the feedback @JEG2 !
You're so encouraging! 

As I advance I'm getting more confident with the code base.
I think we could drop an e-mail at ruby-talk list and do some tweeting about this develop release.
So we could get some attention and some feedback to the next develop releases until we ship the stable 2.0.0.

When we ship the 2.0.0 we can think about spending some time to write some blog posts highlighting the most interesting things of this refactoring process.

**Comment by abinoam** [comment]
Merged! :+1: Closing.

**Comment by abinoam** [comment]
@djberg96 I have reported the findings in 2 issues on rb-readline 

https://github.com/ConnorAtherton/rb-readline/issues/123

https://github.com/ConnorAtherton/rb-readline/issues/124

**Comment by djberg96** [comment]
@abinoam Thank you. And great job getting this change in!

**Comment by djberg96** [comment]
Any chance of a 2.0 prerelease gem?

**Comment by JEG2** [comment]
This is up to @abinoam, but my opinion is that we should do that around the time we feel the changes are mostly complete.

**Comment by abinoam** [comment]
Hi @djberg96,

Perhaps we could just update the README to point that it is possible to run the "prerelease" version by pointing directly to the git repo at Gemfile as in:

```ruby
source "https://rubygems.org"

gem "highline", git: "https://github.com/JEG2/highline.git"
```

What do you (@djberg96 and @JEG2) think about it?

My fear about the prerelase is that people tend to "trust" them!!! No matter how you advise them.
I think we should mature this code just a little more.
I'm not having the time that I wanted to do it, but I'm seeing some free time upcoming in 2 or 3 weeks. So I could do a "sprint" and  accelerate the releases (no promises).

**Comment by JEG2** [comment]
I prefer the README update, yes.

**Comment by djberg96** [comment]
@abinoam This is what prerelease gems are for - to get people to test drive them. If people trust them too much, that's not your problem. And, as you mentioned, they could still point at your repo with bundler, so it's not like depriving people of a prerelease gem will deter the die hards, but it does make it difficult to do standalone testing for small scripts.

**Comment by abinoam** [comment]
@djberg96 I may agree with the first part of what you said (that we're not saving the brave souls from using experimental code in production).

But, could you explain more detailed the part of "make it difficult to do standalone testing for small scripts"? In my superficial thinking about the problem I find it easy to add the "git" option at the Gemfile. Or just clone the repo and do a ```rake install``` on it to have the gem build and installed. Could paint the scenario more clearly so that I can be sure of it?

NOTE: I'm not a native english speaker, so sometimes is hard for me to avoid sounding rude even if I try hard to sound polite. So... if unsure, I'm really interested in your suggestion. Just need to be sure I understand it clearly.

**Comment by djberg96** [comment]
@abinoam No worries, you didn't come off as rude, and I hope I didn't either.

I suppose people could clone and install the repo, sure, so for a power user I guess there's no effective difference (except perhaps having to install whatever dev dependencies you have in the Rakefile). But, that's still more of a nuisance than just "gem install highline --prerelease". :)

But, people don't use what they don't know about. I think (and hope) that creating a prerelease gem is more likely to get more people using it, because they're more likely to spot it with a gem search or gem update. I have no hard proof, or anything, just my instincts telling me it's more likely to be seen.

Anyway, just my 0.02. It's your work, so do as you feel is best.

**Comment by abinoam** [comment]
Hi @djberg96 ,

I think you have convinced me.
```gem install highline --pre``` is really compelling.

I've taken a look at http://guides.rubygems.org/patterns/#prerelease-gems

With a prerelease gem we will have (summing up):
1) To install the stable (1-7-stable) branch:

  - ```gem install highline``` at the console
  - ```gem "highline"``` at Gemfile

2) To install the development (master) branch:

  - ```gem install highline --pre``` at the console
  - ```gem "highline", ">= 2.0.0-develop.1"``` at Gemfile

Confirm?

So I agree with the prerelease gem. Let's see if @JEG2 sees this way too.

**Comment by abinoam** [comment]
I'll be disconnected for some time. Back at night. And thank you @djberg96 .

**Comment by djberg96** [comment]
@abinoam Yep, I think that's pretty much it. Maybe bundler requires a "pre" option of some sort, I don't remember right off, but I -think- what you have there works.

**Comment by JEG2** [comment]
I'm not against the idea.  I would still probably wait until we're a bit closer to done, but this is @abinoam's show and he gets to make the call.

**Comment by abinoam** [comment]
@djberg96 I think bundler guess the "pre" mode by checking any non-numeric character after the digits of the (semantic) version.

> Most of the version specifiers, like >= 1.0, are self-explanatory. The specifier ~> has a special meaning, best shown by example. ~> 2.0.3 is identical to >= 2.0.3 and < 2.1. ~> 2.1 is identical to >= 2.1 and < 3.0. ~> 2.2.beta will match prerelease versions like 2.2.beta.12. 
Source: http://bundler.io/v1.10/gemfile.html

**Comment by abinoam** [comment]
>  but this is @abinoam's show and he gets to make the call.

magic_rabbit_icon_256


@JEG2 you made blush :blush: !!! 

I have a midway solution. 

As code moved from place to place some of the documentation today is completely broken. 

Let me have some time to fix the documentation and then we release the "prerelease" version of the gem. 

What do you think about it?

**Comment by JEG2** [comment]
Sounds perfect to me.

**Comment by abinoam** [comment]
#154

**Comment by djberg96** [comment]
@abinoam How's it looking?

**Comment by abinoam** [comment]
I've had some personal issues that made me late. I think Ill have time to begin by thursday.
(I've already watched some videos of the Yard creator, as I'm not good at documenting)
